### PR TITLE
benchmark put with perceptron

### PR DESCRIPTION
**Pull request by shubham808**
[WIP] we could iterate over this a bit to clean it up :)

### CODE REVIEW

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/features/DotFeatures.h>
+#include <shogun/features/iterators/DotIterator.h>
+#include <shogun/labels/BinaryLabels.h>
+#include <shogun/mathematics/Math.h>
+
+namespace shogun
+{
+	// Mock PerceptronP
+	class PerceptronP : public CSGObject
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: PerceptronP? What does the P stand for?

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		bool train(CDotFeatures* feats, CBinaryLabels* labels, bool use_put)
+		{
+			int32_t num_feat = feats->get_dim_feature_space();
+			SGVector<float64_t> weights(num_feat);
+			weights.set_const(1.0 / num_feat);
+			bias = 0;
+
+			SGVector<float64_t> output(feats->get_num_vectors());
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: i guess the output here can be remove just like in the other PR

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		void TearDown(const ::benchmark::State&)
+		{
+			SG_UNREF(feats);
+			SG_UNREF(labels);
+		}
+
+		void createNormalFeatures(const benchmark::State& state)
+		{
+			index_t num_dim = state.range(0);
+			index_t num_vecs = 1000;
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: can we make this a parameter, both N and D should be parameters

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			SGMatrix<float64_t> mat(num_dim, num_vecs);
+			for (index_t i = 0; i < num_vecs; i++)
+			{
+				for (index_t j = 0; j < num_dim; j++)
+				{
+					mat(j, i) = CMath::normal_random(0.0, 1.0);
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: I think we have some random normal matrix method somewhere ... so no loop is needed

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		void createLabels(const benchmark::State& state)
+		{
+			index_t num_dim = state.range(0);
+			SGVector<float64_t> vec(num_dim);
+
+			for (index_t i = 0; i < num_dim; i++)
+			{
+				vec[i] = CMath::random(0, 1) > 0.5 ? -1 : 1;
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: can be dine in a single line using std::for_each or so

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+			SG_REF(labels);
+		}
+	};
+
+	BENCHMARK_DEFINE_F(DataFixture, Put_perceptron)(benchmark::State& st)
+	{
+		for (auto _ : st)
+		{
+			auto perceptron = new PerceptronP();
+			SG_REF(perceptron);
```
Comment (vigsterkr) [inline src/shogun/util/PutPerceptron_benchmark.cc]: no need for this if you are gonna throw away the obj on the end of the loop

File: src/shogun/util/PutPerceptron_benchmark.cc (hunk 4 above)
Comment (vigsterkr) [reply src/shogun/util/PutPerceptron_benchmark.cc]: plz dont use CMath's random interface for getting randoms

File: src/shogun/util/PutPerceptron_benchmark.cc (hunk 5 above)
Comment (vigsterkr) [reply src/shogun/util/PutPerceptron_benchmark.cc]: same here... dont use random interface of CMath to do this

File: src/shogun/util/PutPerceptron_benchmark.cc (hunk 1 above)
Comment (vigsterkr) [reply src/shogun/util/PutPerceptron_benchmark.cc]: is this really meant to be merged into develop? or rather have it in a feature branch?

**Comment by shubham808** [comment]
These are the results with current code.... i will paste the new results when i update the pr

| Benchmark                         | Time     | CPU      | Iterations |
|-----------------------------------|----------|----------|------------|
| DataFixture/Put_perceptron/8/0    | 162 ms   | 162 ms   | 4          |
| DataFixture/Put_perceptron/64/0   | 184 ms   | 184 ms   | 3          |
| DataFixture/Put_perceptron/512/0  | 884 ms   | 884 ms   | 1          |
| DataFixture/Put_perceptron/4096/0 | 3165 ms  | 3165 ms  | 1          |
| DataFixture/Put_perceptron/8192/0 | 6281 ms  | 6277 ms  | 1          |
| DataFixture/Put_perceptron/8/1    | 658 ms   | 657 ms   | 1          |
| DataFixture/Put_perceptron/64/1   | 683 ms   | 682 ms   | 1          |
| DataFixture/Put_perceptron/512/1  | 1158 ms  | 1156 ms  | 1          |
| DataFixture/Put_perceptron/4096/1 | 5470 ms  | 5462 ms  | 1          |
| DataFixture/Put_perceptron/8192/1 | 10030 ms | 10019 ms | 1          |

**Comment by karlnapf** [comment]
No this is going to develop IMO, and yes of course we need a different name then

**Comment by karlnapf** [comment]
@shubham808 can you help me parsing the output...Put_perceptron/8192/1 means what exactly?

**Comment by shubham808** [comment]
8192 is num_vectors and 1 is the value of use_bias flag

**Comment by karlnapf** [comment]
I dont understand why the put seems to take proportionally longer for larger dimensions. Is there any memory copying on?
@lisitsyn

**Comment by geektoni** [comment]
Since here it seems a good place to do it, I attach also the benchmarking for the `observe()` method.
You can find them here: https://gist.github.com/geektoni/1e01f693b0c479f1c1cf40916445dafd.
As we thought, calling the observe() method and attaching an observer to the machine can slow down the execution by 3000 milliseconds (that's the worst case). 

The code I used to run the test is here: https://github.com/geektoni/shogun/blob/feature/perceptron_observe_benchmark/src/shogun/util/PutPerceptron_benchmark.cc#L143

It is mostly based on Shubham's work in this PR.

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+					if (CMath::sign<float64_t>(predicted_label) != true_label)
+					{
+						const auto gradient = learn_rate * true_label;
+						bias += gradient;
+						v.add(gradient, weights);
+					}
+					if (use_put)
+					{
+						this->put<SGVector<float64_t>>("w", weights);
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: this should be move to outside the loop over the data

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+		virtual const char* get_name() const
+		{
+			return "PerceptronP";
+		}
+
+	protected:
+		/** learning rate */
+		float64_t bias;
+		SGVector<float64_t> w;
+		float64_t learn_rate;
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: i guess we can make learn rate to constant 1 and bias to constant 0 to simplify things

File: src/shogun/util/PutPerceptron_benchmark.cc (hunk 7 above)
Comment (shubham808) [reply src/shogun/util/PutPerceptron_benchmark.cc]: yes! the results after this are:

| Benchmark                         | Time     | CPU      | Iterations |
|-----------------------------------|----------|----------|------------|
| DataFixture/Put_perceptron/8/0    | 120 ms   | 119 ms   | 6          |
| DataFixture/Put_perceptron/64/0   | 193 ms   | 193 ms   | 4          |
| DataFixture/Put_perceptron/512/0  | 837 ms   | 837 ms   | 1          |
| DataFixture/Put_perceptron/4096/0 | 3782 ms  | 3781 ms  | 1          |
| DataFixture/Put_perceptron/8192/0 | 11831 ms | 11731 ms | 1          |
| DataFixture/Put_perceptron/8/1    | 124 ms   | 123 ms   | 6          |
| DataFixture/Put_perceptron/64/1   | 172 ms   | 172 ms   | 4          |
| DataFixture/Put_perceptron/512/1  | 615 ms   | 614 ms   | 1          |
| DataFixture/Put_perceptron/4096/1 | 5285 ms  | 5278 ms  | 1          |
| DataFixture/Put_perceptron/8192/1 | 10290 ms | 10279 ms | 1          |

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+			perceptron->train(feats, labels);
+		}
+	}
+
+	BENCHMARK_DEFINE_F(DataFixture, perceptron_with_put)(benchmark::State& st)
+	{
+		for (auto _ : st)
+		{
+			auto perceptron = new CMockPerceptron();
+			std::function<void()> callback = [&perceptron]() {
```
Comment (shubham808) [inline src/shogun/util/PutPerceptron_benchmark.cc]: @geektoni we can add the observer test similarly now :)

**Comment by karlnapf** [comment]
This seems to be ready? I'll read one more time. Can we have the results here as well?

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+			SG_REF(labels);
+		}
+	};
+
+	BENCHMARK_DEFINE_F(DataFixture, perceptron)(benchmark::State& st)
+	{
+		for (auto _ : st)
+		{
+			auto perceptron = new CMockPerceptron();
+			std::function<void()> callback = [&perceptron]() {
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: i think this is good!

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+			for (index_t i = 0; i < num_dim; i++)
+			{
+				vec[i] = std::rand() % 100 > 50 ? -1 : 1;
+			}
+			labels = new CBinaryLabels(vec);
+			SG_REF(labels);
+		}
+	};
+
+	BENCHMARK_DEFINE_F(DataFixture, perceptron)(benchmark::State& st)
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: maybe call this "perceptron_baseline"

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 12]
```diff
@@ -0,0 +1,149 @@
+#include <benchmark/benchmark.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/features/DotFeatures.h>
+#include <shogun/features/iterators/DotIterator.h>
+#include <shogun/labels/BinaryLabels.h>
+
+namespace shogun
+{
+
+	std::normal_distribution<float64_t> normal;
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: @vigsterkr will have comments here, or? :)

File: src/shogun/util/PutPerceptron_benchmark.cc [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+			std::function<void()> callback = [&perceptron]() {
+				perceptron->put<SGVector<float64_t>>("w", perceptron->weights);
+			};
+			perceptron->m_callback = callback;
+			perceptron->train(feats, labels);
+		}
+	}
+
+	BENCHMARK_REGISTER_F(DataFixture, perceptron)
+	    ->Ranges({{8, 1 << 8}, {8 << 5, 8 << 8}})
```
Comment (karlnapf) [inline src/shogun/util/PutPerceptron_benchmark.cc]: can i suggest to put line breaks in between the Ranges and then add an inline comment on what this addresses? make it easier to read

**Comment by shubham808** [comment]
| Benchmark                                | Time   | CPU    | Iterations |
|------------------------------------------|--------|--------|------------|
| DataFixture/perceptron_baseline/8/256    | 25 ms  | 25 ms  | 28         |
| DataFixture/perceptron_baseline/64/256   | 30 ms  | 30 ms  | 23         |
| DataFixture/perceptron_baseline/256/256  | 61 ms  | 61 ms  | 12         |
| DataFixture/perceptron_baseline/8/512    | 53 ms  | 53 ms  | 14         |
| DataFixture/perceptron_baseline/64/512   | 80 ms  | 80 ms  | 8          |
| DataFixture/perceptron_baseline/256/512  | 130 ms | 130 ms | 5          |
| DataFixture/perceptron_baseline/8/2048   | 208 ms | 208 ms | 3          |
| DataFixture/perceptron_baseline/64/2048  | 238 ms | 238 ms | 3          |
| DataFixture/perceptron_baseline/256/2048 | 965 ms | 965 ms | 1          |

| Benchmark                                | Time   | CPU    | Iterations |
|------------------------------------------|--------|--------|------------|
| DataFixture/perceptron_with_put/8/256    | 35 ms  | 35 ms  | 19         |
| DataFixture/perceptron_with_put/64/256   | 52 ms  | 52 ms  | 12         |
| DataFixture/perceptron_with_put/256/256  | 86 ms  | 86 ms  | 8          |
| DataFixture/perceptron_with_put/8/512    | 71 ms  | 71 ms  | 10         |
| DataFixture/perceptron_with_put/64/512   | 83 ms  | 83 ms  | 7          |
| DataFixture/perceptron_with_put/256/512 | 168 ms |      168 ms |         4
| DataFixture/perceptron_with_put/8/2048        |  254 ms       | 254 ms |         3
| DataFixture/perceptron_with_put/64/2048       |  267 ms       | 267 ms |         3
| DataFixture/perceptron_with_put/256/2048      |  821 ms      |  821 ms |         1

**Comment by karlnapf** [comment]
The difference seems pretty marginal to me.

**Comment by karlnapf** [comment]
Let's merge and then we can put @geektoni stuff in.
If we need changes, we can still do that as it is a feature branch only

**Comment by shubham808** [comment]
@karlnapf yes it is not much of a difference ;)

**Pull request merged**